### PR TITLE
Add capability of timeframe based rotation

### PR DESCRIPTION
**Pull request by r-raymond**
Hi, I've created a small add-on to fast-logger.

It introduces a new type of rotation of log files based on time frames. Any
type of time frame expressible via time format strings can be used; E.g.
using a format string `%d.%m.%Y` will append log files with the current
date and rotate as soon as the day changes. This can be used to create
hourly, minutely, weekly, monthly, etc. rotations.

If in a time frame no log messages appear, that frame will be skipped and
does not count towards the number of files saved

Since there is in general no way to enumerate the log files from these
type of rotations, it keeps track of the former file names for deletion. This
implies that it will not pick up old files upon a restart of the program.

It makes use of the time format caching, just as TimedLogger does. As such
the resolution is restricted similarly, so only time-frames of more than ~10 seconds
make sense.

I've tried to not change any behavior of the fast-logger system and have
therefore always created new data types and functions instead of integrating
with the current. In an ideal world some of the types and functions could be
merged in the future.

Let me know if you have any questions, comments or critique!

### CODE REVIEW

**Comment by kazu-yamamoto** [comment]
I'm playing this PR in https://github.com/kazu-yamamoto/logger/tree/daily-rotate

Travis CI is fixed.

Before merging this PR, I need to switch `time` to `hourglass`.

**Comment by kazu-yamamoto** [comment]
> Before merging this PR, I need to switch `time` to `hourglass`.

Never mid. The following proposal makes `unix-time` cross-flatform: https://github.com/kazu-yamamoto/unix-time/issues/43

I will tackle this time issue independetly.

File: fast-logger/System/Log/FastLogger/File.hs [hunk 1]
```diff
@@ -1,10 +1,21 @@
...
-import Control.Monad (unless, when)
+import Control.Concurrent (forkIO)
+import Control.Monad (unless, when, void)
+import Data.ByteString (ByteString)
+import Data.ByteString.Char8 (unpack)
 import System.Directory (doesFileExist, doesDirectoryExist, getPermissions, writable, renameFile)
-import System.FilePath (takeDirectory)
+import System.FilePath (takeDirectory, dropFileName, takeFileName, (</>))
+
+type FormattedTime = ByteString -- redeclaration to allow for LANGUAGE Safe
```
Comment (kazu-yamamoto) [inline fast-logger/System/Log/FastLogger/File.hs]: This is a duplicated definition. If you agree, I will create `Types.h` to define useful types.

**Comment by buecking** [comment]
just curious, why the move to hourglass? 

In ~2014 i tried swapping in hourglass into a latency sensitive application and it didn't perform well. I'd be curious to see some benchmarks if u know of any.

File: fast-logger/System/Log/FastLogger.hs [hunk 2]
```diff
@@ -259,6 +278,19 @@ newTimedFastLogger tgetter typ = case typ of
                 pushLogStr lgrset (f t)
                 when (cnt <= 0) $ tryRotate lgrset fspec ref mvar
         return (logger, rmLoggerSet lgrset)
+    dailyRotateLoggerInit fspec bsize = do
+        lgrset <- newFileLoggerSet bsize $ daily_log_file fspec
+        cache <- newTimeCache "%Y-%m-%dT%H:%M:%S"
```
Comment (kazu-yamamoto) [inline fast-logger/System/Log/FastLogger.hs]: The format is hard-coded. Can we extend `DailyFileLogSpec` to specify a format?

**Comment by buecking** [comment]
There are a couple of other issues:

the current PR doesn't prefix the date when the initial log is created. 

have not implement a user supplied callback function to process rotated files.

**Comment by kazu-yamamoto** [comment]
> just curious, why the move to hourglass?

In the specific case of `fast-logger`, I would like to remove `#ifdef`.

I don't know `hourglass` is faster than `time` but I can trust `hourglass` simpley because of no floating point conversions.

File: fast-logger/System/Log/FastLogger.hs (hunk 2 above)
Comment (r-raymond) [reply fast-logger/System/Log/FastLogger.hs]: In principal absolutely. The problem is the current implementation (unlike the first) only supports daily rotation. This way I need to be able to figure out when a new day begins, which I cannot in general on a user supplied time format. Another way would be to use yet another time cache internally and one for the user specified format.

**Comment by r-raymond** [comment]
Ok, I think all the before mentioned points have been addressed. It now immediately creates the first log file with the date prefixed. The date(time) is customizable, but the user has to specify herself whether two date(times) are in the same logging time frame. After creation of a new log file, the post_process function gets called on the latest log_file in a different thread.

Let me know if there are any other concerns, comments.

File: fast-logger/System/Log/FastLogger/File.hs (hunk 1 above)
Comment (buecking) [reply fast-logger/System/Log/FastLogger/File.hs]: I don't have permissions to push to `r-raymond/logger.git` but here's a patch
```
diff --git a/fast-logger/System/Log/FastLogger.hs b/fast-logger/System/Log/FastLogger.hs
index 6ecdbcf..8506fbc 100644
--- a/fast-logger/System/Log/FastLogger.hs
+++ b/fast-logger/System/Log/FastLogger.hs
@@ -43,6 +43,8 @@ module System.Log.FastLogger (
   , module System.Log.FastLogger.Date
   -- * File rotation
   , module System.Log.FastLogger.File
+  -- * Types
+  , module System.Log.FastLogger.Types
   ) where
 
 #if __GLASGOW_HASKELL__ < 709
@@ -63,6 +65,7 @@ import System.Log.FastLogger.IORef
 import System.Log.FastLogger.LogStr
 import System.Log.FastLogger.Logger
 import System.Log.FastLogger.Date
+import System.Log.FastLogger.Types
 
 ----------------------------------------------------------------
 
diff --git a/fast-logger/System/Log/FastLogger/Date.hs b/fast-logger/System/Log/FastLogger/Date.hs
index ceb6f5d..d9666d7 100644
--- a/fast-logger/System/Log/FastLogger/Date.hs
+++ b/fast-logger/System/Log/FastLogger/Date.hs
@@ -4,18 +4,15 @@
 -- |
 -- Formatting time is slow.
 -- This package provides mechanisms to cache formatted date.
-module System.Log.FastLogger.Date (
-  -- * Types
-    TimeFormat
-  , FormattedTime
-  -- * Date cacher
-  , newTimeCache
+module System.Log.FastLogger.Date
+  ( -- * Date cacher
+    newTimeCache
   , simpleTimeFormat
   , simpleTimeFormat'
   ) where
 
 import Control.AutoUpdate (mkAutoUpdate, defaultUpdateSettings, updateAction)
-import Data.ByteString (ByteString)
+import System.Log.FastLogger.Types (TimeFormat, FormattedTime)
 #if WINDOWS
 import qualified Data.ByteString.Char8 as BS
 import Data.Time (UTCTime, formatTime, getCurrentTime, utcToLocalZonedTime)
@@ -31,12 +28,6 @@ import System.Posix (EpochTime, epochTime)
 
 ----------------------------------------------------------------
 
--- | Type aliaes for date format and formatted date.
-type FormattedTime = ByteString
-type TimeFormat = ByteString
-
-----------------------------------------------------------------
-
 #if WINDOWS
 -- | Get date using UTC.
 getTime :: IO UTCTime
diff --git a/fast-logger/System/Log/FastLogger/File.hs b/fast-logger/System/Log/FastLogger/File.hs
index 4952e3e..84b9587 100644
--- a/fast-logger/System/Log/FastLogger/File.hs
+++ b/fast-logger/System/Log/FastLogger/File.hs
@@ -13,9 +13,7 @@ import Data.ByteString (ByteString)
 import Data.ByteString.Char8 (unpack)
 import System.Directory (doesFileExist, doesDirectoryExist, getPermissions, writable, renameFile)
 import System.FilePath (takeDirectory, dropFileName, takeFileName, (</>))
-
-type TimeFormat = ByteString -- redeclaration to allow for LANGUAGE Safe
-type FormattedTime = ByteString -- redeclaration to allow for LANGUAGE Safe
+import System.Log.FastLogger.Types (TimeFormat, FormattedTime)
 
 -- | The spec for logging files
 data FileLogSpec = FileLogSpec {
diff --git a/fast-logger/System/Log/FastLogger/Types.hs b/fast-logger/System/Log/FastLogger/Types.hs
new file mode 100644
index 0000000..682b7d0
--- /dev/null
+++ b/fast-logger/System/Log/FastLogger/Types.hs
@@ -0,0 +1,13 @@
+module System.Log.FastLogger.Types (
+  -- * Types
+    TimeFormat
+  , FormattedTime
+  ) where
+
+import Data.ByteString (ByteString)
+
+----------------------------------------------------------------
+  
+-- | Type aliaes for date format and formatted date.
+type FormattedTime = ByteString
+type TimeFormat = ByteString
diff --git a/fast-logger/fast-logger.cabal b/fast-logger/fast-logger.cabal
index 40aa884..8646fd4 100644
--- a/fast-logger/fast-logger.cabal
+++ b/fast-logger/fast-logger.cabal
@@ -17,6 +17,7 @@ Library
   Exposed-Modules:      System.Log.FastLogger
                         System.Log.FastLogger.File
                         System.Log.FastLogger.Date
+                        System.Log.FastLogger.Types
   Other-Modules:        System.Log.FastLogger.IO
                         System.Log.FastLogger.FileIO
                         System.Log.FastLogger.IORef
```

**Comment by buecking** [comment]
Sample usage in it's current state

```
logType = LogFileTimedRotate
     TimedFileLogSpec fname "%FT%H%M%S" sametime compressFile) where
    compressFile fp = void $ readProcess "tar" [ "--remove-files", "-caf", fp <> ".xz", fp ] []
    sametime = (==) `on` C8.takeWhile (/='T')
```

File: fast-logger/System/Log/FastLogger/File.hs (hunk 1 above)
Comment (buecking) [reply fast-logger/System/Log/FastLogger/File.hs]: Got permissions and pushed.

**Comment by kazu-yamamoto** [comment]
@buecking Why do you use `readProcess`? Why not `callProcess`?

**Comment by kazu-yamamoto** [comment]
I will merge https://github.com/kazu-yamamoto/logger/tree/r-raymond-master when we reach consensus on `forkIO` and `readProcess`. 

In this branch:
- a warning is removed
- doc is updated
- ver-bump commit is removed

**Comment by kazu-yamamoto** [comment]
One more question: how about moving `fork` stuff to `timed_post_process`? That is, it is user's responsibility.

**Comment by buecking** [comment]
> @buecking Why do you use `readProcess`? Why not `callProcess`?

This is an example of how a user could use `timed_post_process`.  This should have no affect on fast-logger itself.

**Comment by buecking** [comment]
> One more question: how about moving fork stuff to timed_post_process? That is, it is user's responsibility.

**Comment by kazu-yamamoto** [comment]
Rebased and merged. 
Thank you for your contribution!
If you have any problems, please open a new issue.

**Comment by kazu-yamamoto** [comment]
A new version has been released.

**Pull request closed without merging**